### PR TITLE
Support optional Diagnostic.tags

### PR DESCRIPTION
**Pull request by leungbk**
https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#diagnosticTag

A DiagnosticTag can be either 1 (DiagnosticTag.Unnecessary) or
2 (DiagnosticTag.Deprecated).  Following the rendering suggestions in
the protocol, we fade out Unnecessary code and strike-through
Deprecated code.

* eglot.el (eglot--lsp-interface-alist): Add Diagnostic.tags.
(eglot-client-capabilities): Advertise supported tags.
(eglot-handle-notification): Assign the appropriate properties.

### CODE REVIEW

File: eglot.el [hunk 1]
```diff
@@ -330,6 +330,10 @@ This can be useful when using docker to run a language server.")
     (21 . "Constant") (22 . "Struct") (23 . "Event") (24 . "Operator")
     (25 . "TypeParameter")))
 
+(defconst eglot--tag-face-properties
+  `((1 . (:weight ultra-light))
+    (2 . (:strike-through t))))
```
Comment (skangas) [inline eglot.el:335]: Should these be proper faces instead?

**Comment by skangas** [comment]
Thanks, this looks like a very useful feature. I would suggest that we should have faces instead of a defconst here, to make it easier for users to customize this with the full power of customize-option+defface. WDYT?

**Comment by leungbk** [comment]
> Thanks, this looks like a very useful feature. I would suggest that we should have faces instead of a defconst here, to make it easier for users to customize this with the full power of customize-option+defface. WDYT?

Done.

File: eglot.el [hunk 2]
```diff
@@ -1833,7 +1849,11 @@ COMMAND is a symbol naming the command."
 					     ((<= sev 1) 'eglot-error)
                                              ((= sev 2)  'eglot-warning)
                                              (t          'eglot-note))
-                                       message `((eglot-lsp-diag . ,diag-spec)))))
+                                       message `((eglot-lsp-diag . ,diag-spec))
+                                       (and tags
+                                            `((face . ,(mapcar (lambda (tag)
```
Comment (joaotavora) [inline eglot.el:1854]: Are multiple faces supported here?

Comment (leungbk) [reply eglot.el:1854]: I agree it looks weird, but since Diagnostic.tags is plural, I don't see why not.

Comment (joaotavora) [reply eglot.el:1854]: And do we, meaning, Emacs, support  more than one face in the `cdr` of that list?

Comment (leungbk) [reply eglot.el:1854]: According to 40.9.2 of the Elisp manual, yes:

```
‘face’
     This property controls the appearance of the text (*note Faces::).
     The value of the property can be the following:

        [...]

        • A list of faces.  Each list element should be either a face
          name or an anonymous face.  This specifies a face which is an
          aggregate of the attributes of each of the listed faces.
          Faces occurring earlier in the list have higher priority.
```

Comment (joaotavora) [reply eglot.el:1854]: OK!

**Comment by dgutov** [comment]
FWIW, `company-capf` uses the `:company-deprecated` property for the latter. And has a matching face definition.

The "unnecessary" tag never came up before.

**Comment by joaotavora** [comment]
I generally think this PR is good to go (after the usual git commit tweaks maybe).

@dgutov can you explain what your suggestion is?  How can `:company-deprecated` help here?  AFAIU this is for diagnostics, nor completions

**Comment by leungbk** [comment]
Oh yeah, do we want a test for this before merging? Pyright sends the Unused property, and that was how I checked my commit. I'm not at home right now, but I can get around to it by tomorrow.

**Comment by dgutov** [comment]
Ah, sorry.

I thought you were going to add "deprecated" support for completions, which is also missing in Eglot.

**Comment by skangas** [comment]
@leungbk A test case would be useful, I think.

**Comment by leungbk** [comment]
@skangas Done. My test case (requires `npm i -g pyright`) passes after running `make check`.

**Comment by leungbk** [comment]
Actually...hold on, I need to spend some time debugging the test.

**Comment by leungbk** [comment]
@skangas OK, now it works. I used `rust-analyzer` instead since `pyright-langserver` behaved unexpectedly when running tests.

**Comment by skangas** [comment]
This seems to lead to a byte-compiler warning:
```
eglot.el:1822:1: Warning: Destructuring for Diagnostic has extraneous (:tags)
```

**Comment by joaotavora** [comment]
Nice.  That just means that the Diagnostic type further up in the file must also be changed, informing Eglot that `tags`  is a new optional field.

**Comment by leungbk** [comment]
> This seems to lead to a byte-compiler warning:
> 
> ```
> eglot.el:1822:1: Warning: Destructuring for Diagnostic has extraneous (:tags)
> ```

I can't reproduce this with `make clean && make check` in the root directory. The `Diagnostic` type seems to be appropriately extended with the new optional `:tags` field; you might be looking at one of my earlier (botched) attempts at fixing a merge conflict, when I made a typo extending the `Diagnostic` type.

**Comment by skangas** [comment]
> I can't reproduce this with `make clean && make check` in the root directory.

Hmm, I can reproduce this consistently... but it seems like I can't reproduce this after restarting Emacs, nor can I reproduce it from `emacs -Q`. I guess something was going on with the Emacs session where I'm seeing this. Sorry for the noise.

**Comment by skangas** [comment]
I had to increase the test timeout to 5 seconds, but now everything is working here. I see this diagnostic when replicating the test case manually:
```
rustc: unused variable: `test`
`#[warn(unused_variables)]` on by default
rustc: if this is intentional, prefix it with an underscore: `_test`
```
So I fixed up the commit message slightly (feel free to take a look at the changes) and merged this to master. Thanks for your contribution!

**Comment by theothornhill** [comment]
> > I can't reproduce this with `make clean && make check` in the root directory.
> 
> Hmm, I can reproduce this consistently... but it seems like I can't reproduce this after restarting Emacs, nor can I reproduce it from `emacs -Q`. I guess something was going on with the Emacs session where I'm seeing this. Sorry for the noise.

This can happen if you `eval-buffer` `eglot`. I've seen this sometimes as well, and I believe it has with some of the `eval-and-compile`s running around.

**Comment by joaotavora** [comment]
> This can happen if you eval-buffer eglot. I've seen this sometimes as well, and I believe it has with some of the eval-and-compiles running around.

Yes, Eglot is meant to be compiled, then loaded. If you just interpret it you will get some of that trouble when you do try to compile it.  So I recommend always `M-x emacs-lisp-byte-compile-and-load` when loading the whole file.  Even for individual definitions, I tend to prefer `compile-defun`.  But of course `C-M-x` and `C-x C-e` also work 99% of the time.

Nice test, btw, @leungbk! Seems very clean. And @skangas now that there's a `rust-analyzer` test there, maybe the "watches files" test can also migrate from `rls` to `rust-analyzer`?

**Comment by skangas** [comment]
> @skangas now that there's a `rust-analyzer` test there, maybe the "watches files" test can also migrate from `rls` to `rust-analyzer`?

Yup, I'll look into it.

**Pull request merged**